### PR TITLE
Add shared Module Resources

### PR DESCRIPTION
**Pull request by natebosch**
Extract the behavior of `_CleanMetaModuleCache` into a shared class and
use it consistently when reading in Module or MetaModule instances from
assets.

Any time we read a `Module` or `MetaModule` from a json serialized asset
we consider them to be immutable and safe to share for the duration of a
build. We have a consistent pattern of deserialization and construction.

- Add `DecodingCache` to handle the logic of caching and invalidating
  instances of classes that are serialized to an asset.
- Add a single resource for each of `Module` and `MetaModule` so that
  they can be used consistently.
- Update the places where we were json decoding and constructing these
  classes to all use the shared resources.

### CODE REVIEW

**Comment by natebosch** [comment]
@jakemac53 - one thing that makes me a little nervous is that these classes _aren't_ actually immutable, but we treat them as if they are.

I wonder if we should explore making them immutable for real.

File: build_modules/lib/src/module_cache.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+final metaModuleCache = Resource<DecodingCache<MetaModule>>(
+    () => DecodingCache((m) => MetaModule.fromJson(m)),
+    dispose: (c) => c.dispose());
+
+final moduleCache = Resource<DecodingCache<Module>>(
+    () => DecodingCache((m) => Module.fromJson(m)),
+    dispose: (c) => c.dispose());
+
+class DecodingCache<T> {
```
Comment (jakemac53) [inline build_modules/lib/src/module_cache.dart]: nit: add some doc comments describing this class

File: build_modules/lib/src/module_cache.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  final _cached = <AssetId, Future<Result<T>>>{};
+
+  final T Function(Map<String, dynamic>) _construct;
+
+  DecodingCache(this._construct);
+
+  void dispose() => _cached.clear();
+
+  Future<T> find(AssetId id, AssetReader reader) async {
```
Comment (jakemac53) [inline build_modules/lib/src/module_cache.dart]: document that this returns null if `id` can't be read

File: build_modules/lib/src/module_cache.dart [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    dispose: (c) => c.dispose());
+
+final moduleCache = Resource<DecodingCache<Module>>(
+    () => DecodingCache((m) => Module.fromJson(m)),
+    dispose: (c) => c.dispose());
+
+class DecodingCache<T> {
+  final _cached = <AssetId, Future<Result<T>>>{};
+
+  final T Function(Map<String, dynamic>) _construct;
```
Comment (jakemac53) [inline build_modules/lib/src/module_cache.dart]: nit: `_fromJson`?

File: build_modules/lib/src/module_cache.dart (hunk 1 above)
Comment (natebosch) [reply build_modules/lib/src/module_cache.dart]: Done.

File: build_modules/lib/src/module_cache.dart (hunk 3 above)
Comment (natebosch) [reply build_modules/lib/src/module_cache.dart]: Done.

File: build_modules/lib/src/module_cache.dart (hunk 2 above)
Comment (natebosch) [reply build_modules/lib/src/module_cache.dart]: Done.

File: build_modules/lib/src/module_cache.dart [hunk 4]
```diff
@@ -10,31 +10,46 @@ import 'package:build/build.dart';
...
-  DecodingCache(this._construct);
+  DecodingCache._(this._fromJson);
 
-  void dispose() => _cached.clear();
+  void _dispose() => _cached.clear();
 
+  /// Find and deserialize a [T] stored in [id].
+  ///
+  /// If the asset at [id] is unreadable the returned future will resolve to
+  /// null. If the isntance is cached it will not be decoded again, but the
```
Comment (jakemac53) [inline build_modules/lib/src/module_cache.dart]: nit: `null` and `instance`

**Pull request merged**